sdk/lib/atl/atltrace.h

```diff
@@ -0,0 +1,203 @@
+/*
+ * PROJECT:     ReactOS ATL
+ * LICENSE:     LGPL-2.0-or-later (https://spdx.org/licenses/LGPL-2.0-or-later)
+ * PURPOSE:     Providing ATLTRACE macro
+ * COPYRIGHT:   Copyright 2022 Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>
+ */
+ 
+#pragma once
+
+#include "atldef.h"
+
+#ifdef NDEBUG
+    #undef DBG
+    #undef _DEBUG
+#endif
+
+#if DBG && !defined(_DEBUG)
+    #define _DEBUG
+#endif
+
+#ifdef _DEBUG
+
+#include <stdio.h>
+
+namespace ATL
+{
+
+struct CTraceCategory
+{
+    explicit CTraceCategory(LPCSTR name, UINT level = 0) { }
+
+    operator UINT() const { return (1 << 15); }
+};
+
+#define DEFINE_TRACE_CATEGORY(name, cat) extern const DECLSPEC_SELECTANY UINT name = cat
+DEFINE_TRACE_CATEGORY(atlTraceGeneral,    (1 << 0));
+DEFINE_TRACE_CATEGORY(atlTraceCOM,        (1 << 1));
+DEFINE_TRACE_CATEGORY(atlTraceQI,         (1 << 2));
+DEFINE_TRACE_CATEGORY(atlTraceRegistrar,  (1 << 3));
+DEFINE_TRACE_CATEGORY(atlTraceRefcount,   (1 << 4));
+DEFINE_TRACE_CATEGORY(atlTraceWindowing,  (1 << 5));
+DEFINE_TRACE_CATEGORY(atlTraceControls,   (1 << 6));
+DEFINE_TRACE_CATEGORY(atlTraceHosting,    (1 << 7));
+DEFINE_TRACE_CATEGORY(atlTraceDBClient,   (1 << 8));
+DEFINE_TRACE_CATEGORY(atlTraceDBProvider, (1 << 9));
+DEFINE_TRACE_CATEGORY(atlTraceSnapin,     (1 << 10));
+DEFINE_TRACE_CATEGORY(atlTraceNotImpl,    (1 << 11));
+DEFINE_TRACE_CATEGORY(atlTraceAllocation, (1 << 12));
+#undef DEFINE_TRACE_CATEGORY
+
+struct CTrace
+{
+    enum
+    {
+        DefaultTraceLevel = 0,
+        DisableTracing = 0xFFFFFFFF,
+        EnableAllCategories = 0xFFFFFFFF
+    };
+
+    static UINT GetLevel()      { return s_level; }
+    static UINT GetCategories() { return s_categories; }
+    static void SetLevel(UINT level)           { s_level = level; }
+    static void SetCategories(UINT categories) { s_categories = categories; }
+
+    static bool IsTracingEnabled(UINT category, UINT level)
+    {
+        if (s_level == DisableTracing || s_level < level || !(s_categories & category))
+            return false;
+        return ::IsDebuggerPresent();
```

Comment:
I don;'t think this should depend on the presence of a usermode debugger?

Reply:
Deleted `::IsDebuggerPresent();` call.